### DISCUSSION TITLE
Community fluxes

### DISCUSSION THREAD

**Discussion by beva005** [micom-dev/micom#105] [Q&A] [answered]
Hello,

I am working with MICOM to model a co-culture, but I have doubts on how to define the exchange fluxes. If I use the grow workflow like this:

```pythom
mydatabase = build_database(manifest=df_database,
                             out_path='/home/sara/Desktop/SBV/models/',
                             threads=2,
                             progress=True)
	manifest=build(data,out_folder="/home/sara/Desktop/SBV/models/",model_db=None,cutoff=0.0001, threads=2,solver="cplex")
	com = load_pickle("/home/sara/Desktop/SBV/models/sample_1.pickle")
	medium=	{'reaction': ['EX_CO_m', 'EX_H2_m'],'flux': ['22.0','0'],'metabolite':['CO_m','H2_m']} #Use + ub
	df_medium=pd.DataFrame(medium)
	res = grow(manifest, model_folder="/home/sara/Desktop/SBV/models/", medium=df_medium, tradeoff=0.5, threads=2)

 I get this error:
  File "MICOM_local.py", line 149, in <module>
    res = grow(manifest, model_folder="/home/sara/Desktop/SBV/models/", medium=df_medium, tradeoff=0.5, threads=2)
  File "/home/sara/.local/lib/python3.6/site-packages/micom/workflows/grow.py", line 188, in grow
    results = workflow(_growth, args, threads)
  File "/home/sara/.local/lib/python3.6/site-packages/micom/workflows/core.py", line 55, in workflow
    results = list(it)
  File "/home/sara/.local/lib/python3.6/site-packages/rich/progress.py", line 167, in track
    sequence, total=total, description=description, update_period=update_period
  File "/home/sara/.local/lib/python3.6/site-packages/rich/progress.py", line 1119, in track
    for value in sequence:
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 735, in next
    raise value
TypeError: bad operand type for unary -: 'str'
```

I have also tried constraining directly some uptakes in the models.

```python
	com.optimize()
	sol = com.optimize()
	print(sol)
	print(sol.members)
	sol = com.optimize(fluxes=True, pfba=False)
	for i in sol.fluxes:
		print('{}\t{}\t{}\t{}'.format(i,sol.fluxes[i]['auto'],sol.fluxes[i]['kluy'],sol.fluxes[i]['medium']))
```

But I get weird results. How should I define the exchange reactions?  In my models, the reactions are defined like this EX_XX_e => , so uptake is negative, and production is positive. Are the EX_XX_m reactions defined the other way around? 

Thanks a lot,
Sara

**Comment by cdiener** [comment]
You are very close. The flux values in your medium definition have to be numbers not string. So using 

```python
medium= {'reaction': ['EX_CO_m', 'EX_H2_m'],'flux': [22.0,0],'metabolite':['CO_m','H2_m']} #Use + ub
df_medium=pd.DataFrame(medium)
```

That should run, but you probably will see a growth rate of zero since no organism will be able to grow on CO alone. So you will need a more complete medium.

You don't need to worry about the direction of your exchanges. MICOM figures this out by itself. So the fluxes in the medium are always positive when you specify them.

**Comment by beva005** [comment]
I have just seen the error as well, thanks a lot! 
I have tried to just constrain some uptakes I know in the model itself (e.g EX_CO_e) and let things being uptaken (like doing FBA):
	com.optimize()
	sol = com.optimize()
	#print(sol)
	print(sol.members)
	sol = com.optimize(fluxes=True, pfba=False)
	for i in sol.fluxes:
		print('{}\t{}\t{}\t{}'.format(i,sol.fluxes[i]['auto'],sol.fluxes[i]['kluy'],sol.fluxes[i]['medium']))

I see some metals, phosphate, etc being uptaken, I get some growth in one of the species, but not in the other one. So, there might be some bounds that don't allow flux. 
For example, I constrain this reaction EX_CO_e to -20, but EX_CO_m is -7 and there is no other way CO can be uptaken because the transport has 0 flux.. Shouldn't this one has -20 mmol/gDW/h as well?  Another example, EX_ALA_e has a positive flux in one species and negative in the other species when this reaction has bounds of [0,1000] in both species. I am not sure how it is MICOM changing these bounds.
Can I just constrain these EX_XX_m reactions or better to specify them all in the grow workflow as medium?  

Also, can I change somewhere the total community biomass? 

Thanks,
Sara

**Comment by beva005** [comment, marked as answer]
I have figured out what the issue was!
